### PR TITLE
Implement fetching from Gitlab API

### LINKED ISSUE #4: implement data fetching
- define some types
- ajax it via `purescript-affjax`
- parse into types via `purescript-argonaut` or `purescript-simple-json`
- expose the whole thing as
  - `getProjects :: BaseUrl -> Token -> Aff Projects`
  - `getJobs :: BaseUrl -> Token -> Project -> Aff Jobs`
- all of that could live in a some `Dashboard.Server` module

### PR DESCRIPTION
**Pull request by f-f**
Close #4

### CODE REVIEW

File: src/Gitlab.purs [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+getProjects :: forall a. BaseUrl -> Token -> Aff (ajax :: AJAX | a) Projects
+getProjects baseUrl token = do
+  let url = baseUrl
+            <> "/api/v4/projects?private_token="
+            <> token
+            <> "&simple=true&per_page=20&order_by=last_activity_at"
+  projectsRes <- get url
+  let ps = case readJSON projectsRes.response of
+        Left e -> []
```
Comment (zudov) [inline src/Gitlab.purs]: Please, let's not swallow the errors in such way.

We are working inside `Aff`, error handling (throwing/catching) is a core feature of `Aff` (and it got very cool operations for that, like `apothize`, `attempt`, `invincible`, `supervise`, `bracket`). So throwing errors is morally right thing to do. `get` can throw in a bunch of circumstances (like transport errors), throwing on deserialization failure which can't be handled in any good way and renders the application empty goes very much in line with that.

They document it quite well [here](https://github.com/slamdata/purescript-aff#2-monaderror), just `throwError`.

File: src/Gitlab.purs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+getJobs baseUrl token project = do
+  let url = baseUrl
+            <> "/api/v4/projects/"
+            <> show project.id
+            <> "/jobs?private_token="
+            <> token
+            <> "&per_page=100"
+  jobsRes <- get url
+  let js = case readJSON jobsRes.response of
+        Left e -> []
```
Comment (zudov) [inline src/Gitlab.purs]: Let's `throwError` here as well. At this point you might want to define a helper like `readJsonThrowing :: MonadThrow Error m => ReadForeign a => String -> m a` which would `throwError` an `Error` with `ForeignError`s attached to it.

File: src/Main.purs [hunk 3]
```diff
@@ -2,13 +2,35 @@ module Main where
...
 main = do
-  token     <- URLParams.get "private_token"
-  gitlabUrl <- URLParams.get "gitlab_url"
-  log (token <> " " <> gitlabUrl)
+  token   <- URLParams.get "private_token"
+  baseUrl <- URLParams.get "gitlab_url"
+  log (baseUrl <> " " <> token)
+  launchAff $ do
+    projects <- getProjects baseUrl token
+    liftEff $ log $ writeJSON projects
```
Comment (zudov) [inline src/Main.purs]: If you don't enjoy `liftEff`ing when logging -- there is `Control.Monad.Aff.Console` that totally matches the API of `Control.Monad.Eff.Console` except for one asynchronous nuance.

File: src/Gitlab.purs [hunk 4]
```diff
@@ -0,0 +1,76 @@
+module Gitlab where
+
+import Prelude
+
+import Control.Monad.Aff (Aff)
+import Data.Either (Either(..))
+import Data.Maybe (Maybe(..))
+import Network.HTTP.Affjax (AJAX, get)
+import Simple.JSON (readJSON)
```
Comment (zudov) [inline src/Gitlab.purs]: How about qualifying some of these imports?

Something like:

```purescript
import Network.HTTP.Affjax (AJAX)
import Network.HTTP.Affjax as Ajax
import Simple.JSON as Json
```

File: src/Gitlab.purs [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+type BaseUrl = String
+type Token = String
```
Comment (zudov) [inline src/Gitlab.purs]: `newtype` them if you wish. It's not such a big overhead, all you'll need is to wrap the reading of those in the `main:

```purescript
token   <- Token   <$> URLParams.get "private_token"
baseUrl <- BaseUrl <$> URLParams.get "gitlab_url"
```

and unwrap them in functions that take them:

```purescript
getJobs (BaseUrl baseUrl) (Token token) project =
```

At the expense of a few characters these types will stop being a lie ;)

File: src/Gitlab.purs (hunk 4 above)
Comment (zudov) [reply src/Gitlab.purs]: On the other hand, there are only three functions that are imported from there, so explicitly listing the names and just using them looks quite good too, if we don't expect the amount of imports to SUDDENLY grow.

However `get` is quite prone to name clashes (it already possibly clashes with URLParams.get), so having `Ajax.get` seems very useful and more readable even at this stage.

See yourself. Maybe leave `Simple.JSON` as is, and just make `Network.HTTP.Affjax` qualified as `Ajax`?

**Review by zudov — COMMENTED** [review]
Looks good overall. Could you send me some credentials so I can also give it a run?

File: src/Gitlab.purs (hunk 4 above)
Comment (f-f) [reply src/Gitlab.purs]: No bikeshedding on imports allowed :P 

IMO we should have the policy "if it doesn't import all names from a namespace it's fine".

If/when the imports will suddenly grow we'll fix it.

File: src/Gitlab.purs (hunk 2 above)
Comment (f-f) [reply src/Gitlab.purs]: Do you have a code example how to go about this?

**Comment by f-f** [comment]
@zudov implemented fixes, ready :)

**Comment by zudov** [comment]
Looks good 👍

**Pull request merged**